[FSDP][4/N] Refactor func to share state/init handle attrs

awgu opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #91044 [FSDP][7/N] Support `replicate` in `fully_shard`
* #90959 [FSDP][6/N] Add note explaining idioms for `_FSDPState` traversal
* #90874 [FSDP][5/N] Add manual "wrapping" support for `fully_shard`
* **#90871 [FSDP][4/N] Refactor func to share state/init handle attrs**
* #90945 fully_shard load state_dict

For `limit_all_gathers`, if we do not enforce that they all have the same value, then the entire semantics guaranteed by the `bool` can be violated. It could be as if none of them set that value to be `True`.

For `use_orig_params`, optimizer state dict assumes that the value is the same for all FSDP instances.

Commit:
[FSDP][4/N] Refactor func to share state/init handle attrs

[ghstack-poisoned]

Commit:
Update on "[FSDP][4/N] Refactor func to share state/init handle attrs"




[ghstack-poisoned]

Commit:
Update on "[FSDP][4/N] Refactor func to share state/init handle attrs"




[ghstack-poisoned]

Commit:
Update on "[FSDP][4/N] Refactor func to share state/init handle attrs"




[ghstack-poisoned]

Commit:
Update on "[FSDP][4/N] Refactor func to share state/init handle attrs"




[ghstack-poisoned]

Commit:
Update on "[FSDP][4/N] Refactor func to share state/init handle attrs"




[ghstack-poisoned]

Commit:
Update on "[FSDP][4/N] Refactor func to share state/init handle attrs"




[ghstack-poisoned]

Commit:
Update on "[FSDP][4/N] Refactor func to share state/init handle attrs"




[ghstack-poisoned]

mrshenli commented on torch/distributed/fsdp/_runtime_utils.py:
```diff
@@ -154,60 +162,83 @@ def _lazy_init(
-    return state
+    for attr_name, attr_values in attr_name_to_values.items():
+        if len(attr_values) != 1:
+            raise ValueError(
```
I assume this is for the manual wrapping case, as this will never happen with auto wrap policy?

awgu replied:
Correct!

mrshenli commented on torch/distributed/fsdp/_runtime_utils.py:
```diff
@@ -47,6 +47,14 @@
 
+# Do not include "process_group" to enable hybrid shard and MoE cases
+HOMOGENEOUS_ATTR_NAMES = (
+    "backward_prefetch",
+    "forward_prefetch",
+    "_use_orig_params",
+    "limit_all_gathers",
```
why these needs to be homogeneous? No need to change this, just wonder if there is any hard requirement for this behavior.

awgu replied:
This is a good point. I should have written out the explanation to prove it to myself as well. Refreshing my understanding of the implementation, actually `backward_prefetch` and `forward_prefetch` do not need to be homogeneous. However, I believe the other two do.

For `limit_all_gathers`, if we do not enforce that they all have the same value, then the entire semantics guaranteed by the `bool` can be violated. It could be as if none of them set that value to be `True`.

For `use_orig_params`, optimizer state dict assumes that the value is the same for all FSDP instances.

awgu replied:
I will remove the check for `backward_prefetch` and `forward_prefetch`.

Commit:
Update on "[FSDP][4/N] Refactor func to share state/init handle attrs"




[ghstack-poisoned]

Commit:
Update on "[FSDP][4/N] Refactor func to share state/init handle attrs"


For `limit_all_gathers`, if we do not enforce that they all have the same value, then the entire semantics guaranteed by the `bool` can be violated. It could be as if none of them set that value to be `True`.

For `use_orig_params`, optimizer state dict assumes that the value is the same for all FSDP instances.


[ghstack-poisoned]

Commit:
Update on "[FSDP][4/N] Refactor func to share state/init handle attrs"


For `limit_all_gathers`, if we do not enforce that they all have the same value, then the entire semantics guaranteed by the `bool` can be violated. It could be as if none of them set that value to be `True`.

For `use_orig_params`, optimizer state dict assumes that the value is the same for all FSDP instances.


[ghstack-poisoned]

Commit:
Update on "[FSDP][4/N] Refactor func to share state/init handle attrs"


For `limit_all_gathers`, if we do not enforce that they all have the same value, then the entire semantics guaranteed by the `bool` can be violated. It could be as if none of them set that value to be `True`.

For `use_orig_params`, optimizer state dict assumes that the value is the same for all FSDP instances.


[ghstack-poisoned]

Commit:
Update on "[FSDP][4/N] Refactor func to share state/init handle attrs"


For `limit_all_gathers`, if we do not enforce that they all have the same value, then the entire semantics guaranteed by the `bool` can be violated. It could be as if none of them set that value to be `True`.

For `use_orig_params`, optimizer state dict assumes that the value is the same for all FSDP instances.


[ghstack-poisoned]
